### PR TITLE
Support v4 Legends

### PR DESCRIPTION
**Pull request by domoritz**

### CODE REVIEW

File: src/scale.ts [hunk 1]
```diff
@@ -51,6 +50,9 @@ const SCALE_CATEGORY_INDEX: {
   log: 'numeric',
   pow: 'numeric',
   sqrt: 'numeric',
+  quantize: 'numeric',
```
Comment (kanitw) [inline src/scale.ts]: linear/log/pow (numeric scales) map  Q => Q values

quantile, quantize, threshold map quantitative to discrete values and thus shouldn't be the same scale type. 

If you want to support these, I suggest that you separate adding them to another PR first since there are probably delicate details that requires thorough review.

File: site/docs/encoding/legend.md [hunk 2]
```diff
@@ -89,19 +89,19 @@ To provide themes for all legends, the legends config (`config: {legend: {...}}`
 
 ### General
 
-{% include table.html props="cornerRadius,fillColor,orient,offset,strokeColor,strokeDash,strokeWidth,padding" source="LegendConfig" %}
+{% include table.html props="columns,columnPadding,cornerRadius,fillColor,gridAlign,orient,offset,rowPadding,strokeColor,strokeDash,strokeWidth,padding" source="LegendConfig" %}
```
Comment (kanitw) [inline site/docs/encoding/legend.md]: A lot of these are now repeated.  Better not repeat them but instead say config support all properties except a, b, c...  (and, if applicable, config can also have a few more properties).

File: site/docs/encoding/legend.md [hunk 3]
```diff
@@ -61,7 +61,7 @@ To customize legends, a `legend` object in [an encoding channel's definition](en
 
 {:#properties}
 
-{% include table.html props="type,entryPadding,format,offset,orient,padding,tickCount,title,values,zindex" source="Legend" %}
```
Comment (kanitw) [inline site/docs/encoding/legend.md]: So entryPadding is removed?  (Sounds like a breaking change.)

File: src/compile/legend/properties.ts [hunk 4]
```diff
@@ -15,15 +15,3 @@ export function values(legend: Legend) {
   }
   return vals;
 }
-
-export function type(t: Type, channel: Channel, scaleType: ScaleType): 'gradient' {
```
Comment (kanitw) [inline src/compile/legend/properties.ts]: I'm curious why this should be removed.  You probably have a good reason, but better explain too. :)

File: src/scale.ts (hunk 1 above)
Comment (kanitw) [reply src/scale.ts]: quantile, quantize, threshold (and bin-linear, bin-ordinal) are called discretizing scales in Vega https://vega.github.io/vega/docs/scales/#discretizing

Comment (kanitw) [reply src/scale.ts]: I think we maybe want to leave if bin-linear and bin-ordinal using their keyword though as we probably don't want to allow merging them?

**Comment by domoritz** [comment]
Most of these changes are in #3778. Let's get that in first.

**Comment by Saba9** [comment]
Vega-Lite should be updated to support Vega's`gradientLength` and `gradientThickness` properties. The Vega-Lite docs currently mention analogous `gradientHeight` and `gradientWidth` properties but they don't currently work.

File: site/docs/encoding/legend.md (hunk 3 above)
Comment (domoritz) [reply site/docs/encoding/legend.md]: Yes

File: src/compile/legend/properties.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/legend/properties.ts]: Now, the type is inferred from the scale type and we do not override it anymore.

File: src/legend.ts [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
 
 
-export interface LegendConfig extends VgLegendConfig, VlOnlyGuideConfig {}
+export interface LegendConfig extends LegendConfigNoSignals, VlOnlyGuideConfig {}
 
 /**
  * Properties of a legend or boolean flag for determining whether to show it.
  */
-export interface Legend extends VgLegendBase, Guide {
+export interface Legend extends BaseLegendConfig, Guide {
```
Comment (kanitw) [inline src/legend.ts]: Shouldn't everything down here now should all extend from vega-typings?

A few exception
- `type` -- we want to override "default") 
- `orient` -- (maybe not an exception if the default is not different from Vega)
- `encoding` - VL specific.

File: src/compile/legend/properties.ts [hunk 6]
```diff
@@ -14,14 +11,9 @@ export function values(legend: Legend, fieldDef: FieldDef<string>) {
...
 
-export function type(t: Type, channel: Channel, scaleType: ScaleType): 'gradient' {
-  if (
-      isColorChannel(channel) && (
-        (t === 'quantitative' && !isBinScale(scaleType)) ||
-        (t === 'temporal' && contains<ScaleType>(['time', 'utc'], scaleType))
-      )
-    ) {
-    return 'gradient';
+export function clipHeight(scaleType: ScaleType) {
```
Comment (kanitw) [inline src/compile/legend/properties.ts]: Write a test for this?

File: src/compile/legend/encode.ts [hunk 7]
```diff
@@ -1,34 +1,25 @@
...
+export function symbols(fieldDef: FieldDef<string>, symbolsSpec: any, model: UnitModel, channel: Channel, legendCmp: LegendComponent): SymbolEncodeEntry {
+  if (legendCmp.get('type') === 'gradient') {
     return undefined;
   }
 
   let out = {
     ...applyMarkConfig({}, model, FILL_STROKE_CONFIG),
     ...mixins.color(model)
-  };
+  } as SymbolEncodeEntry;
```
Comment (kanitw) [inline src/compile/legend/encode.ts]: Why do we have to cast here.  Can't we assign type to out? (Add comment if we can't)

File: site/docs/encoding/legend.md [hunk 8]
```diff
@@ -85,25 +107,8 @@ To customize legends, a `legend` object in [an encoding channel's definition](en
...
-### Labels
-
-{% include table.html props="labelAlign,labelBaseline,labelColor,labelFont,labelFontSize,labelLimit,labelOffset,shortTimeLabels" source="LegendConfig" %}
-
-### Symbols
-
-{% include table.html props="entryPadding,symbolColor,symbolType,symbolSize,symbolStrokeWidth" source="LegendConfig" %}
-
-### Title
+To provide themes for all legends, the legends config (`config: {legend: {...}}`) can all [legend properties](#properties) except `direction`, `format`, `orient`, `tickCount`, `title`, `values`, and `zindex`.
```
Comment (kanitw) [inline site/docs/encoding/legend.md]: "can all" -> "support all"

- orient is supported in Vega legend.

- Perhaps worth mentioning that  `direction` has configs based on type (`gradientDirection` and `symbolDirection`).

File: site/docs/encoding/legend.md [hunk 9]
```diff
@@ -61,7 +61,29 @@ To customize legends, a `legend` object in [an encoding channel's definition](en
 
 {:#properties}
 
-{% include table.html props="type,entryPadding,format,offset,orient,padding,tickCount,title,values,zindex" source="Legend" %}
+### General
+
```
Comment (kanitw) [inline site/docs/encoding/legend.md]: Missing `type`?

File: build/vega-lite-schema.json [hunk 10]
```diff
@@ -4854,55 +5022,72 @@
...
+          "type": "number"
         },
-        "gradientHeight": {
-          "description": "The height of the gradient, in pixels.",
-          "minimum": 0,
+        "cornerRadius": {
+          "description": "Corner radius for the full legend.",
           "type": "number"
         },
-        "gradientLabelBaseline": {
```
Comment (kanitw) [inline build/vega-lite-schema.json]: So this is gone?

File: build/vega-lite-schema.json [hunk 11]
```diff
@@ -4854,55 +5022,72 @@
...
+          "description": "The color of the gradient stroke, can be in hex color code or regular color name.\n\n__Default value:__ `\"lightGray\"`.",
           "type": "string"
         },
         "gradientStrokeWidth": {
-          "description": "The width of the gradient stroke, in pixels.",
+          "description": "The width of the gradient stroke, in pixels.\n\n__Default value:__ `0`.",
           "minimum": 0,
           "type": "number"
         },
-        "gradientWidth": {
```
Comment (kanitw) [inline build/vega-lite-schema.json]: So this is gone?

File: build/vega-lite-schema.json [hunk 12]
```diff
@@ -4918,25 +5103,48 @@
...
+            }
+          ],
+          "description": "The strategy to use for resolving overlap of labels in gradient legends. If `false`, no overlap reduction is attempted. If set to `true` (default) or `\"parity\"`, a strategy of removing every other label is used. If set to `\"greedy\"`, a linear scan of the labels is performed, removing any label that overlaps with the last visible label (this often works better for log-scaled axes).\n\n__Default value:__ `true`."
+        },
         "offset": {
-          "description": "The offset, in pixels, by which to displace the legend from the edge of the enclosing group or data rectangle.\n\n__Default value:__  `0`",
+          "description": "The offset in pixels by which to displace the legend from the data rectangle and axes.\n\n__Default value:__ `18`.",
           "type": "number"
         },
-        "orient": {
```
Comment (kanitw) [inline build/vega-lite-schema.json]: orient is supported in Vega Legend config: https://vega.github.io/vega/docs/config/#legends

File: src/legend.ts (hunk 5 above)
Comment (domoritz) [reply src/legend.ts]: We cannot extend and change the type of some.

File: src/compile/legend/properties.ts (hunk 6 above)
Comment (domoritz) [reply src/compile/legend/properties.ts]: This is not used yet.

Comment (domoritz) [reply src/compile/legend/properties.ts]: Done

File: src/compile/legend/encode.ts (hunk 7 above)
Comment (domoritz) [reply src/compile/legend/encode.ts]: Added a fixme

File: site/docs/encoding/legend.md (hunk 8 above)
Comment (domoritz) [reply site/docs/encoding/legend.md]: Done

File: site/docs/encoding/legend.md (hunk 9 above)
Comment (domoritz) [reply site/docs/encoding/legend.md]: Good catch

File: build/vega-lite-schema.json (hunk 10 above)
Comment (domoritz) [reply build/vega-lite-schema.json]: Yes

File: build/vega-lite-schema.json (hunk 11 above)
Comment (domoritz) [reply build/vega-lite-schema.json]: It's been replaced. Yes

File: build/vega-lite-schema.json (hunk 12 above)
Comment (domoritz) [reply build/vega-lite-schema.json]: Oh, this is annoying. Orient is supported in config and legend but it does not support signals.
